src/platform/plugins/shared/discover/public/application/main/components/layout/cascaded_documents/blocks/use_row_header_components.tsx

```diff
@@ -0,0 +1,523 @@
+  );
+
+  return {
+    renderRowActionPopover,
+    togglePopover,
+  };
+};
+
+const textSlotStyles = {
+  textWrapper: css({
+    minWidth: 0,
+    textWrap: 'nowrap',
+  }),
+  textInner: css({
+    overflow: 'hidden',
+    textOverflow: 'ellipsis',
+  }),
+};
+
+export function useEsqlDataCascadeRowHeaderComponents(
+  editorQueryMeta: ESQLStatsQueryMeta,
+  selectedColumns: string[],
+  togglePopover: ReturnType<typeof useEsqlDataCascadeRowActionHelpers>['togglePopover']
+) {
+  const namedColumnsFromQuery = useMemo(() => {
+    return editorQueryMeta.appliedFunctions.map(({ identifier }) => identifier);
+  }, [editorQueryMeta.appliedFunctions]);
+
+  /**
+   * Renders the title part of the row header.
+   */
+  const rowHeaderTitle = useCallback<
+    NonNullable<DataCascadeRowProps<ESQLDataGroupNode, DataTableRecord>['rowHeaderTitleSlot']>
+  >(
+    ({ rowData, nodePath }) => {
+      const rowGroup = nodePath[nodePath.length - 1];
+      const type = editorQueryMeta.groupByFields.find((field) => field.field === rowGroup)?.type;
+
+      if (type && /categorize/i.test(type)) {
+        return (
+          <div data-test-subj={`${rowData.id}-dscCascadeRowTitlePatternCellRenderer`}>
+            {getPatternCellRenderer(
+              // @ts-expect-error - necessary to match the data shape expectation
+              { flattened: rowData },
+              rowGroup,
+              false
+            )}
+          </div>
+        );
+      }
+
+      return (
+        <EuiText size="s">
+          <EuiTextTruncate
+            truncation="end"
+            text={
+              (rowData[rowGroup] ??
+                i18n.translate('discover.esql_data_cascade.row.action.no_value', {
+                  defaultMessage: '(null)',
+                })) as string
+            }
+          >
+            {(truncatedText) => {
+              return <h4 css={textSlotStyles.textInner}>{truncatedText}</h4>;
+            }}
+          </EuiTextTruncate>
+        </EuiText>
+      );
+    },
+    [editorQueryMeta.groupByFields]
+  );
+
+  /**
+   * Renders the meta part of the row header.
+   */
+  const rowHeaderMeta = useCallback<
+    NonNullable<DataCascadeRowProps<ESQLDataGroupNode, DataTableRecord>['rowHeaderMetaSlots']>
+  >(
+    ({ rowData }) =>
+      selectedColumns
+        .map((selectedColumn) => {
+          // only allow grouping columns to be rendered in the meta part of the row header
+          if (namedColumnsFromQuery.indexOf(selectedColumn) < 0) {
+            return null;
+          }
+
+          return (
+            <EuiFlexGroup alignItems="center" gutterSize="s" responsive={false}>
+              <FormattedMessage
+                id="discover.esql_data_cascade.grouping.function"
+                defaultMessage="<bold>{selectedColumn}: </bold><badge>{selectedColumnValue}</badge>"
+                values={{
+                  selectedColumn,
+                  selectedColumnValue: rowData[selectedColumn] as string | number,
```

Comment:
We do this `rowData[column]` casting a few places, but how do we know the value will always be a string (or number in this case)?

Reply:
Should track, but not a blocker for merging while the FF is disabled.